lib/Sema/CSApply.cpp

```diff
@@ -4732,9 +4732,17 @@ namespace {
               KeyPathExpr::Component::forOptionalChain(objectTy, loc));
           break;
         }
-        case KeyPathExpr::Component::Kind::OptionalForce:
+        case KeyPathExpr::Component::Kind::OptionalForce: {
+          // Handle force optional when it is the first component e.g.
+          // \String?.!.count
+          if (resolvedComponents.empty()) {
+            auto loc = origComponent.getLoc();
+            resolvedComponents.push_back(
+                KeyPathExpr::Component::forOptionalForce(componentTy, loc));
```

Comment:
Shouldn't this store the unwrapped type? Could we perhaps tweak `buildKeyPathOptionalForceComponent` to take the optional component type and SourceLoc?

Reply:
> Shouldn't this store the unwrapped type?
As far as I could see `forOptionalForce` expects the optional type to be unwrapped. 

> Could we perhaps tweak buildKeyPathOptionalForceComponent to take the optional component type and SourceLoc?

I think we could but it is used in some other places where I'm not sure this would apply ... but I think we can simplify this because actually for this case if we get here in the first component we can just do this logic here and don't need to actually call `buildKeyPathOptionalForceComponent` because in that case type is always optional and we don't need to handle l-value types which is what `buildKeyPathOptionalForceComponent` does.

Reply:
> As far as I could see `forOptionalForce` expects the optional type to be unwrapped.

Right, but won't `componentTy` be the optional type here?

> I think we could but it is used in some other places where I'm not sure this would apply ... but I think we can simplify this because actually for this case if we get here in the first component we can just do this logic here and don't need to actually call `buildKeyPathOptionalForceComponent` because in that case type is always optional and we don't need to handle l-value types which is what `buildKeyPathOptionalForceComponent` does.

I think it should apply fine to the other cases, but yeah fair enough if you do the unwrapping manually for the root :)

Reply:
> Right, but won't `componentTy` be the optional type here?

Yeah, sorry ... don't know why I wrote unwrapped, it was supposed to be just "optional type" :)
The same way`buildKeyPathOptionalForceComponent` creates the `forOptionalForce` component from the optional type.